<h2>{{alignmentTreeMenuTitle}}</h2>

<h4>
	Unconstrained <em>Alignments</em>
</h4>
<p>
	<em>Alignment</em> objects may be constrained or unconstrained.
	Unconstrained <em>Alignments</em> are used in GLUE projects to store
	the results of any process aimed at identifying homologies between
	nucleotides. This includes classical algorithms such as MUSCLE or MAFFT
	but also manual techniques. In unconstrained <em>Alignments</em> the
	reference coordinate space is purely notional, and not based on any
	particular <em>Sequence</em>. Nucleotide position columns in this
	coordinate space may be added in an unrestricted way in order to
	accommodate any pairwise homology between member <em>Sequences</em>.
</p>
<p>
	The example unconstrained <em>Alignment</em> shown below contains three
	<em>AlignmentMembers</em>. Each <em>AlignmentMember</em> contains
	multiple <em>AlignedSegments</em>; these map between the member
	sequence coordinates and the reference coordinate system. For example a
	block of 7 nucleotides starting at position 5 in Member 3 is mapped to
	reference coordinates [1,7]. The reference coordinate space has been
	expanded to allow for example insertions present in Member 2 at
	locations [8,10] and [24,26] in the reference space, and an insertion
	present in Members 1 and 3 at location [16,18].
</p>

<img width="100%" src="images/unconstrained.png" />
<h4>
	Constrained <em>Alignments</em>
</h4>
<p>
	<em>ReferenceSequence</em> in GLUE can provide a concrete coordinate
	space in which nucleotide data may be interpreted. <em>Alignment</em>
	objects may be constrained to a <em>ReferenceSequence</em>. This
	association is made at the time an <em>Alignment</em> is created and is
	immutable for the lifetime of the <em>Alignment</em>. The nucleotide <em>Sequence</em>
	underlying the constraining <em>ReferenceSequence</em> provides the
	reference coordinate space for the constrained <em>Alignment</em>.
	Therefore, <em>AlignedSegment</em> objects within a constrained <em>Alignment</em>
	propose homology between a nucleotide block on a member <em>Sequence</em>
	and an equal-length block on the constraining <em>ReferenceSequence</em>.
</p>
<p>
	The example constrained <em>Alignment</em> shown below contains the
	same three member <em>Sequences</em>. In this case the <em>Sequence</em>
	underlying Member 1 has also been selected as the constraining <em>ReferenceSequence</em>.
	Therefore, nucleotide columns exist in this <em>Alignment</em>
	precisely for the nucleotide positions which exist in the Member 1 <em>Sequence</em>.
	Consequently, columns are included for insertions present in Member 1
	relative to Member 2 (e.g. [19,21] in the reference space), and
	relative to Member 3 (e.g. [24,26]). However, this alignment does not
	contain columns for insertions present in Member 2 relative to Member 1
	(e.g. between 13 and 14), although the <em>Alignment</em> does record
	the fact that this insertion exists.
</p>
<img width="100%" src="images/constrained.png" />
<p>
	Unconstrained <em>Alignments</em> have the advantage of being able to
	represent the full set of homologies between any pair of member
	sequences, however they must use an artificial coordinate space to
	achieve this. Constrained <em>Alignments</em> use a concrete coordinate
	space but cannot represent homologies within nucleotide columns if
	those columns only exist in insertions relative to the constraining <em>ReferenceSequence</em>.
</p>

<p>
	The <em>AlignedSegment</em> objects within constrained <em>Alignment</em>
	objects may be derived from those within unconstrained <em>Alignment</em>
	objects where both the member and reference sequences are present.
	However, where the member sequences of the constrained <em>Alignment</em>
	are known to be closely related to the <em>ReferenceSequence</em>,
	another possibility exists. In this case the constrained <em>Alignment</em>
	homologies may be computed using a simple pairwise technique between
	the member and reference sequence, for example based on BLAST.
</p>

<h4>Alignment Trees</h4>

Under construction
