<h2>{{alignmentTreeMenuTitle}}</h2>

<p>
	One of the key features of GLUE is its ability to compute, store and
	query multiple sequence alignments, represented by
	<em>Alignment</em> objects in the data model. In GLUE,
	<em>Alignments</em> may be constrained or unconstrained. Unconstrained
	alignments are best for recording homology between a small number of
	distantly-related sequences, whereas constrained alignments are good
	for capturing the homology between a large number of closely-related
	sequences. Constrained alignments may also be assembled into
	<a ng-click="scrollTo('alignmentTrees')">alignment trees</a> which 
	combine homology information together in a phylogenetic data structure. 
</p>

<p>
<ol>
	<li><a ng-click="scrollTo('unconstrainedAlignments')">Unconstrained alignments</a></li>
	<li><a ng-click="scrollTo('constrainedAlignments')">Constrained alignments</a></li>
	<li><a ng-click="scrollTo('alignmentTrees')">Alignment trees</a></li>
</ol>
</p>

<h3 id="unconstrainedAlignments">
	Unconstrained <em>Alignments</em>
</h3>
<p>
	Unconstrained <em>Alignments</em> are used in GLUE projects to store
	the results of any process aimed at identifying homologies between
	nucleotides. This includes classical algorithms such as MUSCLE or MAFFT
	but also manual techniques. In unconstrained <em>Alignments</em> the
	reference coordinate space is purely notional, and not based on any
	particular <em>Sequence</em>. Nucleotide position columns in this
	coordinate space may be added in an unrestricted way in order to
	accommodate any pairwise homology between member <em>Sequences</em>.
</p>
<p>
	The example unconstrained <em>Alignment</em> shown below contains three
	<em>AlignmentMembers</em>. Each <em>AlignmentMember</em> contains
	multiple <em>AlignedSegments</em>; these map between the member
	sequence coordinates and the reference coordinate system. For example a
	block of 7 nucleotides starting at position 5 in Member 3 is mapped to
	reference coordinates [1,7]. The reference coordinate space has been
	expanded to allow for example insertions present in Member 2 at
	locations [8,10] and [24,26] in the reference space, and an insertion
	present in Members 1 and 3 at location [16,18].
</p>

<img width="100%" src="images/unconstrained.png" />
<h3 id="constrainedAlignments">
	Constrained <em>Alignments</em>
</h3>
<p>
	<em>ReferenceSequence</em> in GLUE can provide a concrete coordinate
	space in which nucleotide data may be interpreted. <em>Alignment</em>
	objects may be constrained to a <em>ReferenceSequence</em>. This
	association is made at the time an <em>Alignment</em> is created and is
	immutable for the lifetime of the <em>Alignment</em>. The nucleotide <em>Sequence</em>
	underlying the constraining <em>ReferenceSequence</em> provides the
	reference coordinate space for the constrained <em>Alignment</em>.
	Therefore, <em>AlignedSegment</em> objects within a constrained <em>Alignment</em>
	propose homology between a nucleotide block on a member <em>Sequence</em>
	and an equal-length block on the constraining <em>ReferenceSequence</em>.
</p>
<p>
	The example constrained <em>Alignment</em> shown below contains the
	same three member <em>Sequences</em>. In this case the <em>Sequence</em>
	underlying Member 1 has also been selected as the constraining <em>ReferenceSequence</em>.
	Therefore, nucleotide columns exist in this <em>Alignment</em>
	precisely for the nucleotide positions which exist in the Member 1 <em>Sequence</em>.
	Consequently, columns are included for insertions present in Member 1
	relative to Member 2 (e.g. [19,21] in the reference space), and
	relative to Member 3 (e.g. [24,26]). However, this alignment does not
	contain columns for insertions present in Member 2 relative to Member 1
	(e.g. between 13 and 14), although the <em>Alignment</em> does record
	the fact that this insertion exists.
</p>
<img width="100%" src="images/constrained.png" />
<p>
	Unconstrained <em>Alignments</em> have the advantage of being able to
	represent the full set of homologies between any pair of member
	sequences, however they must use an artificial coordinate space to
	achieve this. Constrained <em>Alignments</em> use a concrete coordinate
	space but cannot represent homologies within nucleotide columns if
	those columns only exist in insertions relative to the constraining <em>ReferenceSequence</em>.
</p>

<p>
	The <em>AlignedSegment</em> objects within constrained <em>Alignment</em>
	objects may be derived from those within unconstrained <em>Alignment</em>
	objects where both the member and reference sequences are present.
	However, where the member sequences of the constrained <em>Alignment</em>
	are known to be closely related to the <em>ReferenceSequence</em>,
	another possibility exists. In this case the constrained <em>Alignment</em>
	homologies may be computed using a simple pairwise technique between
	the member and reference sequence, for example based on BLAST.
</p>

<h3 id="alignmentTrees">Alignment Trees</h3>

<div class="row">
	<div class="col-md-5">
<p>
	GLUE projects have the option of using a structure called an <b>alignment
		tree</b>. This links together alignments in an evolution-oriented way.
	There are often widely recognised phylogenetic clades such as genotypes
	within a set of virus sequences. The structure of the alignment tree
	reflects the phylogenetic relationships between these clades.
</p>
<p>
	An alignment tree is built by creating constrained <em>Alignment</em>
	objects for each of the clades of interest. These <em>Alignments</em>
	become nodes within the tree. Where a parent-child relationship between
	two clades exists within the evolutionary hypothesis, a special
	relational link is introduced between the corresponding pairs of <em>Alignment</em>
	objects.
</p>
<p>
	There is a special condition called the <b>alignment tree invariant</b>
	which is enforced by GLUE when working with alignment trees: If <em>Alignment</em>
	A is a child of <em>Alignment</em> B, the <em>Sequence</em> acting as
	the constraining <em>ReferenceSequence</em> of <em>Alignment</em> A
	must also be a member sequence of <em>Alignment</em> B. In this way, a
	parent <em>Alignment</em> is forced to contain representative member
	sequences from any child <em>Alignments</em>. The object structure of
	an example alignment tree, demonstrating the invariant, is shown in the
	accompanying diagram.
</p>
	</div>
	<div class="col-md-7">
		<img width="100%" src="images/almtTree.png" />
	</div>
</div>

		<p>
		The constrained <em>Alignment</em> at the root represents an entire
		virus species. Two child <em>Alignments</em> represent genotypes 3 and
		4 (clades within the species). Genotype 3 is further subdivided into
		two subtypes, 3a and 3b. Each constrained <em>Alignment</em> has a
		constraining <em>ReferenceSequence</em>. Within each <em>Alignment</em>
		node there are various <em>AlignmentMember</em> objects, each one
		records the pairwise homology between the member <em>Sequence</em> and
		the constraining <em>ReferenceSequence</em>. The alignment tree
		invariant requires for example that the constraining <em>ReferenceSequence</em>
		of subtype 3a is also a member of its parent, genotype 3.
		</p>

		<p>
			There some advantages to using alignment trees in a GLUE project:
			<ol>
				<li>For a number of reasons, it's generally useful to organise virus sequences hierarchically, according to clade.</li>
				<li>Constrained alignments near the tips of the tree can accurately capture homologies between closely-related sequences.</li>
				<li>The alignment tree invariant guarantees that between any two <em>Sequence</em> objects, there is a path of homologies. 
				This facilitates comparisons of distantly related sequences.</li>
			</ol>	
		</p>

<p></p>


