<h2>{{coreSchemaMenuTitle}}</h2>
<div class="row">
	<div class="col-md-6">
		<p>The data within a GLUE project is structured according to the Core Schema; a fixed set of object types and relationships. These bring a
		certain level of evolution-oriented organisation to virus nucleotide
		sequence data and also capture the objects and relationships most commonly
		required to utilise the data. 
		</p>
		<p>
		In the text below, we outline the main GLUE object types, their intended semantics and the relationships between them. 
		GLUE object types are denoted in italicised CamelCase, e.g. <em>FeatureLocation</em>. Some 
		of the main fields and relationships between objects are labelled on the diagrams.
		</p>
		
		<ol>
			<li><a ng-click="scrollTo('ERDs')">GLUE Objects, Fields and Relationships</a></li>
			<li><a ng-click="scrollTo('Sequences')"><em>Sequences</em> and <em>Sources</em></a></li>
			<li><a ng-click="scrollTo('Features')"><em>Features</em></a></li>
			<li><a ng-click="scrollTo('ReferenceSequences')"><em>ReferenceSequences</em> and <em>FeatureLocations</em></a></li>
			<li><a ng-click="scrollTo('Alignments')"><em>Alignments</em> and <em>AlignmentMembers</em></a></em></a></li>
			<li><a ng-click="scrollTo('AlignmentTree')">The GLUE Alignment Tree</a></li>
			<li><a ng-click="scrollTo('Variations')"><em>Variations</em></a></li>
		</ol>
	</div>
	<div class="col-md-6">
		<img width="80%" src="images/projectModel.png" />
	</div>
</div>
<h3 id="ERDs">GLUE Objects, Fields and Relationships</h3>
<div class="row">
	<div class="col-md-6">
		<p>
			Conceptually, GLUE stores data items as objects. These objects include both field values and relationships with other stored objects. 
			The available object types, fields and relationships are defined in the schema. The core schema is available in every GLUE project, 
			but the schema may also be extended on a per-project basis. 
		</p>
		<p>
			In this documentation we use a diagrammatic notation to document different parts of the schema. 
			In the example schema fragment there are three object types: <em>Species</em>, <em>Specimen</em> and <em>Biologist</em>. 
			
		</p>
	</div>
	<div class="col-md-6">
		<img width="100%" src="images/ERD_example.png" />
	</div>
</div>

<h3 id="Sequences"><em>Sequences</em> and <em>Sources</em></h3>
<div class="row">
	<div class="col-md-8">
		<p>
			A set of FASTA virus nucleotide sequences forms the foundation of
			each GLUE project data set. Each GLUE <em>Sequence</em> object is
			assumed to be a section of viral RNA or DNA. Sequences may originate
			from any laboratory sequencing technique as long as FASTA nucleotide
			data is produced. Sequences may also be in silico constructs, such
			ancestral reconstructions, or consensus sequences which are each
			representatives of a larger set. A <em>Sequence</em> object may cover
			the whole of the virus genome, or only a part. 
		</p>
		<p>
			Within a GLUE project, <em>Sequences</em> must be stored within
			<em>Source</em> objects. These exist purely to group sequences; the meaning or usage of
			these groups is left to the project developer.
			<em>Sequences</em> are identified by their 
			<em>sequenceID</em> string field. However, the uniqueness of the 
			<em>sequenceID</em> value is limited to the containing <em>Source</em> object.
			So, no two <em>Sequences</em> within the same <em>Source</em> may have the same 
			<em>sequenceID</em>; conversely two <em>Sequences</em> within two different <em>Sources</em>
			may have the same <em>sequenceID</em>. 
		</p>
		<p>
			<em>Sequences</em> may be imported into GLUE from multiple file formats. The 
			specific format is recorded in a field value on the <em>Sequence</em> object; 
			the original file data is stored in the database in byte form. The aim of this is to 
			allow retrieval of any non-nucleotide data which may be present in the original data file.
			This original data is immutable for the lifetime of the <em>Sequence</em> object.
		</p>
	</div>
	<div class="col-md-4">
		<img width="50%" src="images/sourceSequence.png" />
	</div>
</div>

<h3 id="Features"><em>Features</em></h3>
<div class="row">
	<div class="col-md-8">
		<p>
		For many virus species, biological properties of certain named regions
		of the virus genome have been suggested or established. Within a GLUE
		project, <em>Feature</em> objects may be created for any genomic region of
		interest. Examples include regions which are understood to code for
		polyproteins or individual proteins, or regions which code for domains
		within a protein. <em>Feature</em> objects may also represent non-coding regions
		such as promoter regions. The <em>Feature</em> objects in a GLUE project may be
		arranged in a hierarchy, with each <em>Feature</em> optionally having a single
		parent <em>Feature</em> object and several child <em>Feature</em> objects associated. This
		should reflect the containment relationships of the corresponding genome
		regions. One example of this would be a genome region that is translated into a precursor 
		polyprotein which is later cleaved into several final proteins. The precursor polyprotein would 
		be defined as one genome feature. Each of the final proteins would be defined as an additional
		feature which is a child of the precursor polyprotein.  
		</p>
		<p>
		<em>Feature</em> objects may also have <em>FeatureMetatag</em> objects
		associated. These are simple metadata properties, with fixed semantics,
		which allow specialised GLUE functionality to interpret <em>Feature</em> objects,
		for example to perform protein translation.
		</p>
	</div>
	<div class="col-md-4">
		<img width="70%" src="images/features.png" />
	</div>
</div>

<h3 id="ReferenceSequences"><em>ReferenceSequences</em> and <em>FeatureLocations</em></h3>

<div class="row">
	<div class="col-md-6">

<p>
GLUE uses <em>ReferenceSequence</em> objects to organise, link and interpret the
wider set of <em>Sequence</em> objects. A <em>ReferenceSequence</em> is based on a
specific <em>Sequence</em> and this relationship is mandatory and immutable for
the lifetime of the <em>ReferenceSequence</em>. The choice of which <em>Sequence</em>
objects to use for <em>ReferenceSequence</em> objects will vary from project to
project, some examples are:
<ul>
	<li>Sequences which have been curated as reference sequences by
		another database, e.g. NCBI RefSeq</li>
	<li>Sequences derived from well-known virus strains which were
		identified early in the study of a particular virus species or subtype</li>
	<li>Sequences derived from virus strains which have become
		established as having important laboratory uses</li>
	<li>Sequences where the genome is particularly well-annotated</li>
	<li>Less frequently, sequences derived from an <em>in silico</em> procedure such as
		computing the consensus of a wider set, or ancestral reconstruction</li>
</ul>
</p>

<p>
One of the main functions of <em>ReferenceSequences</em> in GLUE is to map the specific 
co-ordinates of of genome features. To achieve this, a <em>ReferenceSequence</em> object
may contain <em>FeatureLocation</em> objects in order
to represent the presence and position of specific <em>Features</em>. A <em>FeatureLocation</em>
is considered a wholly contained constituent part of a <em>ReferenceSequence</em>, but it 
is also immutably associated with a single <em>Feature</em> during its lifetime. 
</p>
<p>
A <em>FeatureLocation</em> contains a set of <em>FeatureSegment</em>
objects which specify the coordinates of the associated <em>Feature</em> on the
<em>ReferenceSequence</em>. The <em>FeatureSegment</em> objects contain start and end
points, and together may specify a discontiguous region.
</p>
<p>
Typically, mulitple <em>ReferenceSequences</em> will contain <em>FeatureLocations</em> for 
the same <em>Feature</em>, but with different co-ordinates specified by their <em>FeatureSegments</em> as necessary. 
Additionally, a certain <em>Feature</em> may or may not be present (i.e. be associated with a 
<em>FeatureLocation</em>) on a given
<em>ReferenceSequence</em> in the GLUE project. This flexibility may be useful
for example when specific virus genes exist only in a subset of taxa within the
project scope. 
</p>
</div>
	<div class="col-md-6">
		<img width="100%" src="images/referenceSequences.png" />
	</div>
</div>


<h3 id="Alignments"><em>Alignments</em> and <em>AlignmentMembers</em></h3>

<div class="row">
	<div class="col-md-6">
<p>GLUE projects capture nucleotide homologies between <em>Sequences</em>. At a
	fine-grained level, these homologies may be thought of as statements
	proposing that a certain block of nucleotides in one <em>Sequence</em> has the
	same evolutionary origin as a certain block in a second <em>Sequence</em>.</p>

<p><em>Alignments</em> in GLUE aggregate homologies relating to a set
	of “member” <em>Sequences</em> that are associated with the <em>Alignment</em> via
	<em>AlignmentMember</em> objects. So in concrete terms an <em>Alignment</em>
	contains a set of <em>AlignmentMember</em> objects and each <em>AlignmentMember</em>
	associates a member <em>Sequence</em> with the containing <em>Alignment</em>.</p>

<p>Each <em>Alignment</em> defines a reference
	coordinate space. Homologies between members of the alignment are expressed in terms of this coordinate space. 
	Each <em>AlignmentMember</em> contains a set of
	<em>AlignedSegment</em> objects; these are homology statments locating a block of member <em>Sequence</em> nucleotides in the reference 
	coordinate space. An <em>AlignedSegment</em> is a tuple with
	four integer fields:</p>
<p>
<em>&lt;refStart, refEnd, memberStart, memberEnd&gt;</em>
</p>
<p>
	This states that the block [<em>memberStart</em>, <em>memberEnd</em>]
	in the member <em>Sequence</em> is homologous with the block [<em>refStart</em>,
	<em>refEnd</em>] in the reference coordinate space of the containing
	<em>Alignment</em>. GLUE ensures that these two regions
	are always of equal length.
</p>
<p>This reference coordinate space associated with an <em>Alignment</em>
	object is used as an indirect step relating the member sequence
	nucleotides with each other. So, two blocks of nucleotides from
	distinct member <em>Sequence</em> objects within an <em>Alignment</em> are considered to
	be homologous when they are mapped via <em>AlignedSegment</em> objects to the
	same reference coordinates.</p>
	</div>
	<div class="col-md-6">
		<img width="100%" src="images/alignments.png" />
	</div>
</div>	
<div class="row">
	<div class="col-md-12">
	<h4>Unconstrained <em>Alignments</em></h4>
<p><em>Alignment</em> objects may be constrained or unconstrained.
	Unconstrained <em>Alignments</em> are used in GLUE projects to store the results
	of any process aimed at identifying
	homologies between nucleotides. This includes classical algorithms such
	as MUSCLE or MAFFT but also manual techniques. In
	unconstrained <em>Alignments</em> the reference coordinate space is purely
	notional, and not based on any particular <em>Sequence</em>. Nucleotide position
	columns in this coordinate space may be added in an unrestricted way
	in order to accommodate any pairwise homology between member <em>Sequences</em>.</p>
		<p>
			The example unconstrained <em>Alignment</em> shown below contains
			three <em>AlignmentMembers</em>. Each <em>AlignmentMember</em>
			contains multiple <em>AlignedSegments</em>; these map between the
			member sequence coordinates and the reference coordinate system. For
			example a block of 7 nucleotides starting at position 5 in Member 3
			is mapped to reference coordinates [1,7]. The reference coordinate
			space has been expanded to allow for example insertions present in
			Member 2 at locations [8,10] and [24,26] in the reference space, and
			an insertion present in Members 1 and 3 at location [16,18].
		</p>

		<img width="100%" src="images/unconstrained.png" />
	<h4>Constrained <em>Alignments</em></h4>
<p><em>ReferenceSequence</em> in GLUE can provide a
	concrete coordinate space in which nucleotide data may be interpreted. 
	<em>Alignment</em> objects may be constrained to a <em>ReferenceSequence</em>. 
	This association is made at the time an <em>Alignment</em> is created and is 
	immutable for the lifetime of the <em>Alignment</em>. 
	The nucleotide <em>Sequence</em> underlying the constraining
	<em>ReferenceSequence</em> provides the reference coordinate space for the
	constrained <em>Alignment</em>. Therefore, <em>AlignedSegment</em> objects within a
	constrained <em>Alignment</em> propose homology between a nucleotide block on
	a member <em>Sequence</em> and an equal-length block on the constraining
	<em>ReferenceSequence</em>.</p>
	<p>The example constrained <em>Alignment</em> shown below contains the same three member <em>Sequences</em>.
	In this case the <em>Sequence</em> underlying Member 1 has also been selected as the constraining 
	<em>ReferenceSequence</em>. Therefore, nucleotide 
	columns exist in this <em>Alignment</em> precisely for the nucleotide positions which exist in the Member 1 <em>Sequence</em>.
	Consequently, columns are included for insertions present in Member 1 relative to Member 2 
	(e.g. [19,21] in the reference space), and relative to Member 3 (e.g. [24,26]). However, this alignment 
	does not contain columns for insertions present in Member 2 relative to Member 1 (e.g. between 13 and 14), 
	although the <em>Alignment</em> does record the fact that this insertion exists.</p>
			<img width="100%" src="images/constrained.png" />
<p>Unconstrained <em>Alignments</em> have the advantage of being able to represent the full set of homologies between 
any pair of member sequences, however they must use an artificial coordinate space to achieve this. 
Constrained <em>Alignments</em> use a concrete coordinate space but cannot represent homologies within nucleotide columns if those columns 
only exist in insertions relative to the constraining <em>ReferenceSequence</em>. This limitation is addressed by the <a ng-click="scrollTo('AlignmentTree')">GLUE Alignment Tree</a> concept. </p>
			
<h3 id="AlignmentTree"><em>The GLUE Alignment Tree</em></h3>

<p>The <em>AlignedSegment</em> objects within constrained <em>Alignment</em> objects
	may be derived from those within unconstrained <em>Alignment</em> objects where
	both the member and reference sequences are present. However, where the
	member sequences of the constrained <em>Alignment</em> are known to be closely
	related to the <em>ReferenceSequence</em>, another possibility exists. In this
	case the constrained <em>Alignment</em> homologies may be computed using a
	simple pairwise technique between the member and reference sequence,
	for example based on BLAST.</p>
	</div>
</div>



<h3 id="Variations"><em>Variations</em></h3>