<h2>HCV-GLUE</h2>

	<table class="table">
		<tr>
<td>
<h4>Overview</h4>
The emergence of clinically-relevant hepatitis C virus (HCV) resistance to direct-acting 
antiviral drugs in parallel with advances in next generation sequencing
technology means there is a need for computational resources that allow clinicians to 
make optimal treatment decisions based on analysis of virus sequence data. 
These resources are also an important component of efforts to study and monitor HCV genetic variation, 
and to evaluate its impact on disease course and treatment outcome. 

HCV-GLUE is interactive online framework for rapid, standardised analysis of HCV sequence data.

<h4>Data sources</h4>
Data relating to structural, evolutionary and phenotypic aspects of HCV were obtained 
from various sources. These are integrated together in <a href="../gluetoolsApp">GLUE</a>, a data-centric software 
framework for capturing, developing and utilising nucleotide-level virus evolutionary 
hypotheses. The result is a GLUE project specific to HCV.
</td>
<td>
<img alt="HCV-GLUE data sources" width="700" src="images/web_diag_1.png">
</td>
</tr>
</table>

<h4>HCV-GLUE project structure</h4>

The data objects relating to the virus and the fine-grained relationships between these 
objects are captured in the GLUE relational model. A fragment of the model is shown.

<img alt="HCV-GLUE project structure" width="1100" src="images/web_diag_2.png">


<h4>Web analysis tool</h4>

The HCV-GLUE project is embedded into an <a href="#/analysisTool">interactive sequence analysis tool</a> which operates over the world wide web.

<img alt="HCV-GLUE project structure" width="1100" src="images/web_diag_3.png">

<h4>References</h4>

<ol>
<li>D.B. Smith, J. Bukh, C. Kuiken, A. Scott Muerhoff, C.M. Rice, A.T. Stapleton, and P. Simmonds (2014), 
“Expanded Classification of Hepatitis C Virus Into 7 Genotypes and 67 Subtypes: Updated Criteria and Genotype Assignment Web Resource”, 
HEPATOLOGY, 59:318-327</li>
<li>C. Kuiken, C. Combet, J. Bukh, T. Shin-I, G. Deleage, M. Mizokami, R. Richardson, E. Sablon, K. Yusim, J-M. Pawlotsky, P. Simmonds, and the Los Alamos HIV database group (2006), 
“A Comprehensive System for Consistent Numbering of HCV Sequences, Proteins and Epitopes”, 
HEPATOLOGY, 44:1355-1361</li>
<li>S.F. Altschul, T.L. Madden, A.A. Schäffer, J. Zhang, Z. Zhang, W. Miller, and D.J. Lipman (1997), 
"Gapped BLAST and PSI-BLAST: a new generation of protein database search programs", 
Nucleic Acids Res., 25:3389-3402</li>
</ol>

<h4>Collaborators</h4>

	<table class="table">
		<tr>
			<td>
				<a href="http://www.stop-hcv.ox.ac.uk/">MRC-Stop HCV Consortium</a>
			</td>
			<td>
				<img align="middle" alt="STOP-HCV logo" src="images/STOP-HCV-logo-with-strapline.gif">
			</td>
		</tr>	
		<tr>
			<td>
				<a href="http://www.hcvresearchuk.org/">HCV Research UK</a>
			</td>
			<td>
				<img align="middle" alt="HCV Research UK logo" src="images/hcv_research_uk_logo.png">		
			</td>
		</tr>
		<tr>
			<td>
				<a href="http://www.gla.ac.uk/researchinstitutes/iii/cvr/">MRC-University of Glasgow Centre for Virus Research</a>
			</td>
			<td>
				<img align="middle" alt="CVR logo" src="images/CVRLogo.png">		
			</td>
		</tr>
	</table>

